### PR TITLE
Improve useEventListener types and add React ref support

### PR DESCRIPTION
**Pull request by aaronccasanova**
### WHAT is this pull request doing?

This PR improves the `useEventListener` types by adding inference for `event targets`/`event names` and validating the defined `event name` exists on the `event target`:

https://user-images.githubusercontent.com/32409546/174123714-952a3d7f-5bde-4eff-ad7b-29928175e315.mov

> Additionally: This PR address an issue with the current implementation where state variables were not updating as expected.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/playground/Playground.tsx [hunk 1]
```diff
@@ -1,11 +1,32 @@
 import React from 'react';
 
 import {Page} from '../src';
+import {useEventListener} from '../src/utilities/use-event-listener';
 
 export function Playground() {
+  const ref = React.useRef<HTMLParagraphElement>(null);
```
Comment (aaronccasanova) [inline polaris-react/playground/Playground.tsx]: Reminder to remove these examples before merging

**Review by sam-b-rose — APPROVED** [review]
Love this improvement! Great DX 👏

**Comment by dleroux** [comment]
🙌  Thank you @aaronccasanova it worked: https://github.com/Shopify/web/pull/69580

**Review by dleroux — APPROVED** [review]
Don't forget to remove the playground :) 🙇‍♂️

**Comment by dleroux** [comment]
I can confirm that this 100% working here: https://github.com/Shopify/web/pull/69580

**Comment by dleroux** [comment]
My PR will be required for the Polaris-bump in web

**Comment by alex-page** [comment]
@dleroux mate, can you confirm 110%?

**Comment by dleroux** [comment]
> @dleroux mate, can you confirm 110%?

~~Yeah.~~ https://github.com/Shopify/web/pull/69580 is happy.

We will need my PR before shipping to web because of my comment in my PR:


The new event listener in polaris makes use of optimization method like useRef. Because of this, event Listeners attached to hidden elements like popovers do not register. The workaround to this is [this new hook](https://github.com/Shopify/web/pull/69580/files#diff-49970d11a40dc55c22267e80dee41d8866df545f44a5c062fcb54ea53da18158)] use-element-ref.ts, which use a callback and state to return the node. See [React docs](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node)

**Comment by dleroux** [comment]
@alex-page I'm thinking if we want to be 100% we test all the areas in the admin where this could be an issue.  I'll highlight those in my PR.

**Comment by dleroux** [comment]
> @alex-page I'm thinking if we want to be 100% we test all the areas in the admin where this could be an issue. I'll highlight those in my PR.

You making me paranoid mate :) It's actually all good. There'll attached to window which is no issue. The one that was is fixed.

**Comment by alex-page** [comment]
Maaaaate. You got this. @dleroux

**Pull request merged**